text/0017-execution-results.md

```diff
@@ -0,0 +1,91 @@
+- Proposal Name: Execution Outcome
+- Start Date: 2019-09-23
+- NEP PR: [nearprotocol/neps#0017](https://github.com/nearprotocol/neps/pull/17)
+- Issue(s): https://github.com/nearprotocol/nearcore/issues/1307
+
+# Summary
+[summary]: #summary
+
+Refactor current TransactionResult/TransactionLog/FinalTransactionResult to improve naming, deduplicate results and provide
+results resolution by the front-end for async-calls.
+
+# Motivation
+[motivation]: #motivation
+
+Right now the contract calls 2 promises and doesn't return a value, the front-end will return one of the promises results as an execution result. It's because we return the last result from final transaction result. With the current API, it's impossible to know what is the actual result of the contract execution.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here is the proposed Rust structures. Highlights:
+- Rename `TransactionResult` to `ExecutionOutcome` since it's used for transactions and receipts
+- Rename `TransactionStatus` and merge it with result into `ExecutionResult`.
+- In case of success `ExecutionResult` can either be a value of a receipt_id. This helps to resolve the
+  actual returned value by the transaction from async calls, e.g. `A->B->A->C` should return result from `C`.
+  Also in distinguish result in case of forks, e.g. `A` calls `B` and calls `C`, but returns a result from `B`.
+  Currently there is no way to know.
+- Rename `TransactionLog` to `ExecutionOutcomeWithId` which is `ExecutionOutcome` with receipt_id
+  or transaction hash. Probably needs a better name.
+- Rename `FinalTransactionResult` to `FinalExecutionOutcome`.
+- Update `FinalTransactionStatus` to `FinalExecutionResult`.
+- Provide final resolved returned result directly, so the front-end doesn't need to traverse the receipt tree.
+  We may also expose the error directly in the execution result.
+- Split into final outcome into transaction and receipts.
+
+```rust
+pub struct ExecutionOutcome {
+    /// Execution result.
+    pub result: ExecutionResult,
+    /// Logs from this transaction or receipt.
+    pub logs: Vec<LogEntry>,
+    /// Receipt IDs generated by this transaction or receipt.
+    pub receipt_ids: Vec<CryptoHash>,
+    /// The amount of the gas burnt by the given transaction or receipt.
+    pub gas_burnt: Gas,
+}
+
+pub enum ExecutionResult {
+    /// The execution of the given transaction or receipt is pending.
+    Pending,
```

Comment:
I feel ExecutionStatus/FinalExecutionStatus is still a better name then *Result. Since pending here and Not Started/Started in FinalExecutionResult is not the result but more like a status that will make progress overtime

Reply:
Yeah, we can use `Status`. I just don't want to have both `result` and `status`